### ISSUE TITLE
Performance degrading a lot with high number of keys

### ISSUE THREAD

**Issue by crystalin** [paritytech/parity-db#212]
Running the storage benchmark:
```
/home/ubuntu/projects/moonbeam/target/release/moonbeam
   benchmark
   storage
   --db=rocksdb
   --state-version=0
   --mul=1.1
   --weight-path  /home/ubuntu/projects/moonbeam/weights-rocksdb-alphanet.rs
   --chain alphanet
   --base-path /var/lib/rocksdb-alphanet-data
```

for each chain

Alphanet (~20M keys):
```
pub const RocksDbWeight: RuntimeDbWeight = RuntimeDbWeight {
  read: 65_167 * constants::WEIGHT_REF_TIME_PER_NANOS,
  write: 114_721 * constants::WEIGHT_REF_TIME_PER_NANOS,
};
pub const ParityDbWeight: RuntimeDbWeight = RuntimeDbWeight {
  read: 16_290 * constants::WEIGHT_REF_TIME_PER_NANOS,
  write: 65_374 * constants::WEIGHT_REF_TIME_PER_NANOS,
};
```

Moonriver (~30M keys state):
```
pub const RocksDbWeight: RuntimeDbWeight = RuntimeDbWeight {
  read: 66_865 * constants::WEIGHT_REF_TIME_PER_NANOS,
  write: 114_947 * constants::WEIGHT_REF_TIME_PER_NANOS,
};
pub const ParityDbWeight: RuntimeDbWeight = RuntimeDbWeight {
  read: 14_483 * constants::WEIGHT_REF_TIME_PER_NANOS,
  write: 64_545 * constants::WEIGHT_REF_TIME_PER_NANOS,
};
```

Moonbeam (~110M keys state):
```
pub const RocksDbWeight: RuntimeDbWeight = RuntimeDbWeight {
  read: 33_439 * constants::WEIGHT_REF_TIME_PER_NANOS,
  write: 86_828 * constants::WEIGHT_REF_TIME_PER_NANOS,
};
pub const ParityDbWeight: RuntimeDbWeight = RuntimeDbWeight {
  read: 177_320 * constants::WEIGHT_REF_TIME_PER_NANOS,
  write: 69_450 * constants::WEIGHT_REF_TIME_PER_NANOS,
};
```

**Comment by crystalin** [comment]
Additionally, it took something like 40h to generate the moonbeam storage benchmark

**Comment by ggwpez** [comment]
@cheme do you have an idea why the ParityDB time for read on the 110M keys DB is so much slower than Rocks when it is normally faster?

**Comment by crystalin** [comment]
You can download a recent Moonbeam state: https://s3.console.aws.amazon.com/s3/object/alan-stuff?region=us-east-1&prefix=moonbeam-state-3631095.json.lz4 (10GB) if you want to check it

**Comment by cheme** [comment]
That is definitely not expected. Could imagine worst access on big mmap memory, but not something in these proportions. Can also think of the data not being correctly build (there is a reindexing running in background every N values, but this is flushed on exit/start).
 
" based on substrate 0.9.40." : is it substrate version (looks old)?
Would be interesting to have the parity-db version listed in the Cargo.lock (a version from a few month ago did have an issue that could explan some bad behavior cc\ @arkpar ).

**Comment by crystalin** [comment]
You can access the cargo.lock there: https://raw.githubusercontent.com/PureStake/moonbeam/82035b77cd48c2fffb44907ce1501b2128117213/Cargo.lock
`version = "0.4.6"`

**Comment by crystalin** [comment]
I'm happy to cherry-pick some changes on top of it if you want to test few things. You can also probably reproduce by using the snapshot I provided

**Comment by arkpar** [comment]
I've tried using warp sync on moonbeam. The sync went fine, although peak RAM usage was over 130GB. However the parachain is not finalizing blocks. Final block is still at zero. Is this a known issue? Unfinalized blocks are stored differently in the DB and this may affect performance.

**Comment by arkpar** [comment]
As for possible performance issues, it could be affected by how the benchmark is implemented. RocksDB uses its own caching, while ParityDb relies on the OS cache. IIRC the benchmark warmup touches a few of the keys, and for RocksDB this causes a lot more data to be pre-cached.

**Comment by crystalin** [comment]
@arkpar warp is not fully supported yet. We are still working on it.
I also suspect the benchmark implementation is the reason for those unexpected values, but it is hard/long to verify

**Comment by crystalin** [comment]
@arkpar were you able to reproduce? Let me know if I can help otherwise

**Comment by arkpar** [comment]
I could not access the snapshot linked above. It requires AWS registration and asks for my credit card number. I've started regular sync instead and it looks like it will take 3-4 days.

**Comment by arkpar** [comment]
@crystalin Could you give it a test with parity-db 0.4.10?
`cargo update -p parity-db` should do it

**Comment by crystalin** [comment]
I'm running it now.
This time I looked at the CPU load and IO load, and during the benchmark:
- IOPS: ~1300 (max is 3000)
- CPU: 2.5%
- Memory: 10% (max: 32Gb)

**Comment by ggwpez** [comment]
If the DB benchmark time is a major problem then we could add a flag to only read 10% or 1% of the total keys. That way you would have some preliminary results for faster iterating. Do you think that would help?

**Comment by crystalin** [comment]
That could make sense yes a % flag

**Comment by crystalin** [comment]
Warmup round just finished, I might get result this WE

**Comment by crystalin** [comment]
I was able to run it (with substrate 0.9.43 and paritydb 0.4.10). It took 3 days to finish:
```
pub const ParityDbWeight: RuntimeDbWeight = RuntimeDbWeight {
  read: 182_722 * constants::WEIGHT_REF_TIME_PER_NANOS,
  write: 60_176 * constants::WEIGHT_REF_TIME_PER_NANOS,
};
```

**Comment by cheme** [comment]
I did check a bit more how to switch the chainspec loading to something that do not load all state in memory, but it is a bit more work than I did expect (break a lot the genesis build api since we need to do multiple commit while using a streaming json parser), so I postpone doing this myself for now.
Still I got a better understanding of the benchmarking process and it just use the standard chainspec loading, which means that the full state is send in parity db but the bench run on a db that just got a lot of key injected.
So the db may still be doing one or two levels of table reindexing when doing its benchmark, which would explain the performance issue.

This can be check by doing "ls" on the db directory and looking at the file for the state column:
if it is still reindexing the state there will multiple file named `paritydb/full/index_01_xx` with xx being the index sizes. 

If this is the case I do not have of a simple way of ensuring reindexing (changing default index size to paritydb can be a a hacky solution).

The following change in substrate would allow flushing the logs but would not force all reindexing to finish.
```
--- a/bin/node/cli/src/command.rs
+++ b/bin/node/cli/src/command.rs
@@ -127,6 +127,8 @@ pub fn run() -> Result<()> {
                                        ),
                                        #[cfg(feature = "runtime-benchmarks")]
                                        BenchmarkCmd::Storage(cmd) => {
+                                               // load once first to ensure db is flushed.
+                                               new_partial(&config)?;
                                                // ensure that we keep the task manager alive
                                                let partial = new_partial(&config)?;
                                                let db = partial.backend.expose_db();
```
but it would need to keep db open for a while until everything is reindex too.

Maybe simply doing the bench in two steps:
- step 1 load chainspec (eg just start the bin with no connection to ensure only chainspec loading will progress, or have a specific command to do so). Then wait until there is no more reindexing in paritydb (single index_01_xx file) until exiting.
- step 2 run benchmark on existing db.

Or implement a primitive that ensure all reindexing is finished in paritydb and use it before calling new_partial a second time (but it will not be very elegant as the code at this level do not assume a specific db).

**Comment by crystalin** [comment]
Thank you,

I think we did run the node with no connection (we often do for other profiling parts) before running the benchmark, but I can try again to see if that helps.

I think having substrate support the storage benchmark on a substate of the state would probably be more effective in that case.

**Comment by ggwpez** [comment]
Yes I hope to get https://github.com/paritytech/substrate/issues/14659 to some newcomer to solve.

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/on-ethereum-and-evm-compatibility/13663/12
